clang/lib/CodeGen/Targets/HLSLTargetInfo.h

```diff
@@ -0,0 +1,39 @@
+//===- HLSLTargetInfo.h ---------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "ABIInfoImpl.h"
+#include "TargetInfo.h"
+
+using namespace clang;
+using namespace clang::CodeGen;
+
+//===----------------------------------------------------------------------===//
+// Target codegen info implementation common between DirectX and SPIR/SPIR-V.
+//===----------------------------------------------------------------------===//
+
+class CommonHLSLTargetCodeGenInfo : public TargetCodeGenInfo {
+public:
+  CommonHLSLTargetCodeGenInfo(std::unique_ptr<ABIInfo> Info)
+      : TargetCodeGenInfo(std::move(Info)) {}
+
+  // Returns LLVM target extension type "dx.Layout" or "spv.Layout"
+  // for given structure type and layout data. The first number in
+  // the Layout is the size followed by offsets for each struct element.
+  virtual llvm::Type *getHLSLLayoutType(CodeGenModule &CGM,
+                                        llvm::StructType *LayoutStructTy,
+                                        SmallVector<unsigned> Layout) const {
+    return nullptr;

```

Comment:
Do we need a todo here? Maybe an issue tracking that this needs an implementation?

Reply:
This method is/will be implemented by the derived classes `DirectXTargetCodeGenInfo` and `CommonSPIRTargetCodeGenInfo`. The default implementation should return `nullptr`, the same as the default implementation of `getHLSLType` on the `TargetCodeGenInfo` class. `nullptr` means "I don't know who to convert this type".

Reply:
Should this be a pure virtual function? Deriving from CommonHLSLTargetCodeGenInfo but not defining this function seems like it would be a hard error.

Reply:
I have moved the buffer layout calculations to a separate `HLSLBufferLayoutBuilder` class. This method is no longer needed.